src/wp-includes/block-template.php

```diff
@@ -0,0 +1,228 @@
+ * @return WP_Block_Template|null template A template object, or null if none could be found.
+ */
+function resolve_block_template( $template_type, $template_hierarchy ) {
+	if ( ! $template_type ) {
+		return null;
+	}
+
+	if ( empty( $template_hierarchy ) ) {
+		$template_hierarchy = array( $template_type );
+	}
+
+	$slugs = array_map(
+		'_strip_php_suffix',
+		$template_hierarchy
+	);
+
+	// Find all potential templates 'wp_template' post matching the hierarchy.
+	$query     = array(
+		'theme'    => wp_get_theme()->get_stylesheet(),
+		'slug__in' => $slugs,
+	);
+	$templates = get_block_templates( $query );
+
+	// Order these templates per slug priority.
+	// Build map of template slugs to their priority in the current hierarchy.
+	$slug_priorities = array_flip( $slugs );
+
+	usort(
+		$templates,
+		function ( $template_a, $template_b ) use ( $slug_priorities ) {
+			return $slug_priorities[ $template_a->slug ] - $slug_priorities[ $template_b->slug ];
+		}
+	);
+
+	return count( $templates ) ? $templates[0] : null;
+}
+
+/**
+ * Displays title tag with content, regardless of whether theme has title-tag support.
+ *
+ * @access private
+ * @since 5.8.0
+ *
+ * @see _wp_render_title_tag()
+ */
+function _block_template_render_title_tag() {
+	echo '<title>' . wp_get_document_title() . '</title>' . "\n";
+}
+
+/**
+ * Returns the markup for the current template.
+ *
+ * @access private
+ * @since 5.8.0
+ *
+ * @return string block tempate markup.
+ */
+function get_the_block_template_html() {
+	global $_wp_current_template_content;
+	global $wp_embed;
+
+	if ( ! $_wp_current_template_content ) {
+		if ( is_user_logged_in() ) {
+			return '<h1>' . esc_html__( 'No matching template found' ) . '</h1>';
+		}
+		return;
+	}
+
+	$content = $wp_embed->run_shortcode( $_wp_current_template_content );
+	$content = $wp_embed->autoembed( $content );
+	$content = do_blocks( $content );
+	$content = wptexturize( $content );
+	if ( function_exists( 'wp_filter_content_tags' ) ) {
+		$content = wp_filter_content_tags( $content );
+	} else {
+		$content = wp_make_content_images_responsive( $content );
+	}
+	$content = str_replace( ']]>', ']]&gt;', $content );
+
+	// Wrap block template in .wp-site-blocks to allow for specific descendant styles
+	// (e.g. `.wp-site-blocks > *`).
+	return '<div class="wp-site-blocks">' . $content . '</div>';
+}
+
+/**
+ * Renders a 'viewport' meta tag.
+ *
+ * @access private
+ * @since 5.8.0
+ *
+ * This is hooked into {@see 'wp_head'} to decouple its output from the default template canvas.
+ */
+function _block_template_viewport_meta_tag() {
+	echo '<meta name="viewport" content="width=device-width, initial-scale=1" />' . "\n";
+}
+
+/**
+ * Strips .php suffix from template file names.
+ *
+ * @access private
+ * @since 5.8.0
+ *
+ * @param string $template_file Template file name.
+ * @return string Template file name without extension.
+ */
+function _strip_php_suffix( $template_file ) {
```

Comment:
The naming isn't accurate here, since the line below also removes `.html` suffixes. We should either change the name, or the implementation. Since I only see one callsite which seems `.php` specific, it might be okay to only remove `.php`?

Reply:
It's internal anyway, your call

Reply:
Ah, I think the `.html` case is needed for custom page block templates. Renaming to `_strip_template_file_suffix` in  bf8d81ef8930352d505fa491fcc6a517e47e3f94.